Bugfix and improvement to get_template_cartesian_coordinates

emichr opened the pull request:
- Fixed a bug where only positive x and y template coordinates were selected when a window size was given.

- Added support for interpreting the window size as a radius rather than a rectangular window

- Updated docstring to reflect changes

- Reformatted using Black

---
name: Bugfix and improvement to pyxem.utils.polar_transform_utils.get_template_cartesian_coordinates
about: Fixing a bug where only positive coordinates were returned and added a feature where the window size can be interpreted as a radius rather than a rectangular region

---

**Checklist**
- [ ] Updated CHANGELOG.md
- [ ] (if finished) Requested a review (from pc494 if you are unsure who is most suitable)

**What does this PR do? Please describe and/or link to an open issue.**
This PR corrects an issue with how template coordinates are returned in cartesian coordinates by correcting a condition statment. It also adds a try/except statement where the window size is first attempted to be interpreted as a tuple-like object, and if that fails it is instead interpreted as a number.

**Describe alternatives you've considered**
Instead of a try/except statement, a typing test could be considered. However, that is less less flexible if, for instance a numpy array or list is passed.

**Are there any known issues? Do you need help?**
I don't know if there are any tests related to this function, or if a test should be created.

hakonanes commented:
There are four failing tests, one of which I interpret to be related to this PR: [`test_get_template_cartesian_coordinates()`](https://github.com/pyxem/pyxem/runs/6569934472?check_suite_focus=true#step:8:282). It would be good if you could make sure this test passes locally before pushing. One can run a single test with `pytest -k test_get_template_cartesian_coordinates`.

emichr commented:
The failed test was due to my suggested change not dividing the window size by a factor 2. I.e. I treated the window size parameter as a half-width. I have changed the code accordingly passes the test. One thing I would like feedback on however, is whether it best makes sense to interpret the window size as a radius or a diameter when given as a single number. A diameter would fall more within the "window" approach, but a radius is often more intuitive when it comes to diffraction patterns.

I also added a new tests to test that the window returns correct reflections:
- a window size of (8, 8) should return the mock simulation reflection at (3, 4).
- a window size of 3 should not return any reflection (number is interpreted as a radius)
- a window size of $\sqrt{3^2+4^2}$ should return the reflection at (3, 4)

I considered adding reflections to test the windows after rotating by angles less than 90 degrees, as a square window (given as a tuple) would then behave differently than a circular window (given as a number). However, this requires more advanced testing which I am not sure is desired?

din14970 commented:
> Fixed a bug where only positive x and y template coordinates were selected when a window size was given.

The main reason I built this function was to overlay a template onto the corresponding image in a plot. This is really the only reason you ever need pixel coordinates for the spots. The idea is you align the spots onto the image DP, then crop everything out that is not within the bounds of the image (reason for the weird window definition). If you change this to be a symmetric window around the central spot, it changes the intended behavior of the function, because the real diffraction pattern will not be perfectly centered, so you may have spots "stickout out" from the edges. You will never have negative coordinates because coordinates in an image are all >0. 

That said, it is not entirely clear that this is what the function *should* do. For it to make any sense, the window size should always be the size of the image, which is ambiguous. I could also imagine scenarios where you might rather cut out a symmetric window or radius from the central spot. So I'm not sure at the moment what the best solution is but I think a more substantial refactor may be in order.

din14970 commented on pyxem/utils/polar_transform_utils.py:
```diff
@@ -160,7 +159,17 @@ def get_template_cartesian_coordinates(
     y = s * ox + c * oy + center[1]
     if window_size is not None:
-        condition = (x < window_size[0]) & (y < window_size[1]) & (y >= 0) & (x >= 0)
+        try:
```
For me this has some code smells. Personally I would think it would be more clear to add a different argument, e.g. `limiting_radius`. Then write some boolean logic that determines the precedence of different "filtering" arguments

emichr replied:
Sure, that is probably clearer.

The reason I preferred this solution was that I don't have an overview of how other parts of pyxem that aren't neccesarily included in the tests use this function. Introducing an additional argument could break stuff I didn't want to tamper too much with.

emichr replied:
I'm not sure if any boolean logic is required to determine the precedence of the different "filtering arguments"? If a limiting radius is given that is smaller than the specified window, then you probably wouldn't want to include them anyway, and vica verca. Filtering the reflections first using a limiting radius and then filtering based on the window should do the trick, right?

emichr commented:
Ah, that makes sense! I thought it was used for limiting the template reflections to e.g. the ones that were used in the matching. I sometimes have libraries that extend to e.g the edge of the experimental DPs, but I only do matching on smaller g-vectors. In such cases, I wanted to use the window parameter to this function could be used to limit the visualised reflections to the ones that were actually used in the matching.

As this is not a bug after all, it might make more sense to update the docstring to clarify the behaviour as a first step. Looking at the code for this function should probably be done as part of a larger effort as you say @din14970.

din14970 commented:
> I sometimes have libraries that extend to e.g the edge of the experimental DPs, but I only do matching on smaller g-vectors. 

In principle during the simulation of the library, a maximum radius of simulation can be defined. But you are right that it can not be done only for visualisation. And simulating a library can take a long time. I agree it should be simpler to manipulate these things. I think some work is necessary on diffsims side (also have some ideas to substantially speed up template library simulation), I just haven't gotten to it.
